### PR TITLE
autogram readme

### PR DESCRIPTION
**Pull request by ValerianRey**
- **Implement vgp**
- **Add second version**
- **Change columns to really be columns**
- **Replace vgp by get_output_and_gramian**
- **Use vmap**
- **Fix shape of identity matrix given to vgp**
- **Revert "Fix shape of identity matrix given to vgp"**
- **Add check that the function outputs a vector**
- **Fix dim call**
- **Implements vgp function**
- **Implements get_gramian from vgp_fn**
- **Add has_aux to vgp**
- **"fix" typing of vgp**
- **Add vgp_from_module**
- **rename functional_model_call_v2 to functional_model_call**
- **Make autogram protected**
- **Fix import**
- **Fix device of eye**
- **Fix call to get_gramian**
- **Add @torch.compile around vgp_fn**
- **Revert "Add @torch.compile around vgp_fn"**
- **Remove has_aux option from vgp**
- **[WIP] Add test_algo_3.py**
- **[WIP] Remove vmap, add autojac comparison, etc**
- **Use block diagonal structure efficiently**
- **Add test_sparse_jac.py**
- **Add optimization**
- **Clean up code**
- **Add Timer context manager**
- **Remove dead code**
- **Improve Timer**
- **Remove test_sparse_jac.py**
- **Fix device issues**
- **Add test_reused_tensor**
- **Add test_reused_tensor_autograd**
- **Remove use_deterministic_algorithms**
- **Improve syncing, change bs, change chunk_size of autogram**
- **Add profiling of algo3 vs autojac**
- **Remove tests for reused tensors**
- **Remove profiling script**
- **Refactor**
- **Use curried functions to create tensors on the right device**
- **Fix batching in get_vjp**
- **Use Sequential in Sequential in Cifar10Model**
- **Add test_sequential (autograd chained manually vs all at once)**
- **Revert "Add test_sequential (autograd chained manually vs all at once)"**
- **Rename test_algo3 to test_speed and create test_equivalence**
- **Try backpack**
- **Revert "Try backpack"**
- **Fix weighting instantiation in test_equivalence**
- **add test_per_sample_grads.py**
- **Revert "add test_per_sample_grads.py"**
- **Add AutogramWrapper**
- **Revert "Add AutogramWrapper"**
- **Add multi-threading solution**
- **Revert "Add multi-threading solution"**
- **Add for-loops around speed tests**
- **Add SmartFlatten**
- **Use addmm_**
- **WIP add autogram_forward_backward based on augment_model**
- **Add plot_graph utility**
- **Simplify plot_graph.py**
- **Add `next_edges`**
- **Add `targets_to_leaf_targets`**
- **Simplify targets_to_leaf_targets, rename variables**
- **Add FlatNonSequentialNN in plot_graph.py**
- **Add FlatNonSequentialNN to tests**
- **Improve speed testing**
- **Improve plot_graph**
- **Update plot_graph.py**
- **Add `GramianAccumulator` class**
- **WIP add multi-input multi-output support**
- **WIP add multi-input multi-output support**
- **Fix multi-input and add test for it**
- **Simplify code and add next steps**
- **WIP add autograd function for autogram**
- **Add new implementation with autograd.Function (now working more or less)**
- **Create `get_model_hook`**
- **Create `make_autogram_function`**
- **Cleaning**
- **Detach graph of input in JacobianAccumulator**
- **WIP**
- **Fix multiple output**
- **Remove torchviz**
- **Add more parametrizations to test_speed**
- **Add PyTreeModel**
- **WIP converting to PyTree**
- **remove `outputs` parameter to `make_jacobian_accumulator`**
- **improve usage of `_vjp_from_module` by being careful about detupling.**
- **move reshaping of jacobian in `GramianAccumulator`**
- **remove a TODO**
- **select only one output (the one with fewest elements) as a target for the first backward phase**
- **remove main**
- **Handles the case when a module has no output.**
- **add TODO**
- **Remove useless for comprehension to check if an iterator has an element**
- **Improve clarity of JacobianAccumulator code**
- **Remove SmartFlatten (not useful thanks to the unsqueeze(0) trick)**
- **Add ModuleThatTakesString**
- **Improve todos**
- **Add ModuleWithParameterReuse test**
- **Add ModelWithInterModuleParameterReuse**
- **Add ModelWithModuleReuse**
- **Update todo**
- **Improve condition on having module parameters**
- **Add ModelWithFreeParameter and ModelWithNoFreeParameter tests**
- **Improve some test robustness**
- **Add ModuleWithUnusedParam test**
- **Add ModuleWithFrozenParam test**
- **Add support for requires_grad=False parameters**
- **Fix bug of missing recurse=False**
- **Remove todo**
- **Add test_equivalence_gramian**
- **Some possible interface**
- **Fix bug with prepended hook**
- **Some possible interface**
- **Few things:**
- **Fix multiple calls to augment_model_with_iwrm_autogram**
- **Remove some dosctring as `grad_output` do matter**
- **Implement AutogramActivator**
- **Make the backward phase alternate, we can now in principle backward several time the same tensor.**
- **Remove test_equivalence2 (it was actually testing exactly the same thing as test_equivalence)**
- **Add INPUT_SIZE field to models**
- **Add batch_size parameter to tests**
- **Add resnet18**
- **Add ResNet18 class**
- **Add ModuleWithBuffer**
- **Make assert_tensor_dicts_are_close generic on the dicts keys**
- **Update test_equivalence to allow multi-iteration**
- **Add test_augment_deaugment_reaugment**
- **Refactor test optimization functions to take a loss_fn**
- **Use forward_pass in autojac_get_gramian**
- **Add note about importing from protected _pytree module**
- **Add OUTPUT_SIZE to models, make it possible to test multi-input and multi-output models**
- **Add else clause to targets_to_leaf_targets**
- **Improve condition in gramian property**
- **Add excluded edges to `targets_to_leaf_targets`**
- **Exclude input from `targets_to_leaf_targets` search.**
- **Make test_speed use architecture param instead of model**
- **Add model deaugmentation in test_speed**
- **Move models to _architectures.py**
- **Remove _vgp.py and its tests**
- **Fix bug of trying to get gradient edge of tensors that dont require grad in model_hook**
- **Add ShapedModule**
- **Add casting of W to Weighting[PSDMatrix]**
- **Remove autojac_get_gramian**
- **Remove TODOs**
- **Fix formatting**
- **Make take PyTree args instead of Any.**
- **Add tests with PyTree input**
- **Make Cifar10Model have sequential instead of being sequential**
- **Add test for partial autogram**
- **Fix typing errors**
- **Use `PyTree` for `inputs` in `_vjp_from_module`**
- **Make fields of `GramianAccumulator` private**
- **Make fields of `_ModelAugmenter` private**
- **deactivate mypy on `plot_graph`**
- **Add test with empty output and no effective parameters**
- **Remove unused `activated` parameter for `AutogramActivator`**
- **Replace exception when not all jacobian are accounted for by an assert for code coverage.**
- **fix assertion**
- **Fix test `ModelWithModuleWithoutOutput` and check if the output is empty**
- **Move `plot_graph` to `tests/plots`**
- **Revamp `GramianAccumulator`: * Make private * Document * add `add_jacobians` and adapt usage * Improve naming**
- **Move `_GramianAccumulator` to `_utils.py`**
- **Move `next_edges` to `_utils.py`**
- **Move `targets_to_leaf_targets` to `_utils.py`**
- **Move `_vjp_from_module` to `_utils.py`**
- **Make `GramianAccumulator` public (it is in private file `_utils.py`)**
- **Make `vjp_from_module` public (it is in private file `_utils.py`)**
- **Document `next_edges`**
- **Document `targets_to_leaf_targets`**
- **Document `vjp_from_module`**
- **Add `TargetRegistry`**
- **Add `get_instance_wise_vjp` in `_utils.py`: * make `vjp_from_module` private**
- **Make `activated` a class field in `JacobianAccumulator`. Note that each `JacobianAccumulator` is a different class.**
- **remove one level of nesting in `JacobianAccumulator.backward`**
- **explicit the return type of `_vjp_from_module`.**
- **Fix typing of `get_instance_wise_vjp` and `get_vjp` within it**
- **Add docstring of `get_instance_wise_vjp`**
- **Add `HandleManager`**
- **Add a subclass of `HandleManager` to avoid publishing the method `add_handle` to the user (we prefer publishing only `remove`).**
- **Rename _total_gramian to _gramian in GramianAccumulator**
- **Rename _full_jacobians to _summed_jacobians in GramianAccumulator**
- **Improve type hints in GramianAccumulator.__init__**
- **Make gramian property able to return None**
- **Add return type hints to __init__ in AutogramHandleManager and TargetRegistry**
- **Fix docstring of get_leaf_target_edges**
- **Make next_edges protected**
- **Split _utils.py into several files**
- **Reorder functions in _rev_gram_acc.py**
- **Make make_jacobian_accumulator protected**
- **Rename _rev_gram_acc to _augment_model.py**
- **Add augment_model_with_iwrm_autogram import in _autogram/__init__.py**
- **Add augment_model_with_iwrm_autogram import in main __init__.py**
- **Improve HandleManager.remove docstring**
- **Make HandleManager public**
- **Improve documentation**
- **Fix formatting of index.rst**
- **Add draft explanation of Gramian-based JD in index.rst**
- **Add test_autogram.py in doc tests**
- **Rename augment_model_with_iwrm_autogram to augment_model_for_gramian_based_iwrm**
- **Add `HookActivator` and make `GramianAccumulator` and `TargetRegistry` have a reset method, so that a caller can reset their state without having to be in the same class.**
- **Move `HookActivator` to its own file**
- **Move `_make_autogram_activator` out of `_ModelAugmneter`. In my opinion, this is a good step away from local minimum of having everything in the same class, but this needs some more polishing.**
- **Move autograd.Functions to `_autograd_functions`**
- **Move hooks creation out of `_ModelAugmenter` to `_forward_hooks.py`**
- **Fix condition for module with no output.**
- **Revert "Fix condition for module with no output."**
- **Rename get to make for hooks**
- **Add intermediate steps when registering hooks**
- **Reorder code in _autograd_functions.py**
- **Remove redefinition of type hints**
- **Remove _ModelAugmenter**
- **Add comments in augment_model_for_gramian_based_iwrm**
- **Rename many things related to registry**
- **Rename AutogramActivator to AutogramScaler**
- **Rename autograd.Functions to be capitalized (because they are classes and that's what they do in torch documentation)**
- **Remove assignment of variable**
- **Rename HookActivator to Activator, and state to is_active**
- **Add support for slow tests**
- **Make FlatNonSequentialNN smaller (to make its test faster)**
- **Remove some parametrizations of test_speed**
- **Reduce the number of iterations in test_equivalence to 3**
- **Rename test_algo_3.py to test_augment_model.py**
- **Improve condition for empty output module.**
- **Add docstrings for autograd functions**
- **Add docstrings for Activator**
- **Improve docstring of `augment_model_for_gramian_based_iwrm`**
- **Add dosctring for the forward hooks.**
- **Improve docstring of _make_model_hook**
- **Improve docstring of _make_module_hook**
- **Turn make_module_hook and make_model_hook public**
- **Fix typo in AutogramScaler docstring**
- **Improve JacobianAccumulator docstring**
- **Add docstring to Cifar10Model**
- **Add docstring to FlatNonSequentialNN**
- **rename FlatNonSequentialNN to BranchedModel**
- **Add OverlyNestedModel**
- **Remove ModuleThatTakesString and ModelThatTakesString**
- **Fix input shapes in MultiInputMultiOutputNN and MultiInputNN**
- **Revamp multi-input and multi-output modules/models**
- **Reorder and rename BranchedModel**
- **Rename PyTreeModule to PyTreeOutputModule**
- **Add docstring to PyTreeOutputModule**
- **Improve PyTreeInputPyTreeOutputModule**
- **Add docstring to PyTreeInputPyTreeOutputModule**
- **Rename some models**
- **Add docstring to PIPOBranchedModel**
- **Add docstring to SIPOBranchedModel**
- **Reorder branched models**
- **Add PyTreeInputSingleOutputModule**
- **Add PISOBranchedModel**
- **Reorder EmptyOutputModule**
- **Reorder real architectures**
- **Add docstring to ResNet18 and rename it InstanceNormResNet18**
- **Reorder MatMulModule and add docstring to it**
- **Add docstring to ModuleWithParameterReuse**
- **Rename ModuleWithParameterReuse to ParamReuseModule**
- **Rename ModelWithInterModuleParameterReuse to InterModuleParamReuseModel**
- **Add docstring to InterModuleParamReuseModel**
- **Rename MatMulModule to _MatMulModule**
- **Rename ModelWithModuleReuse to ModuleReuseModel and add docstring**
- **Make ModuleReuseModel and ModelWithModuleWithoutOutput not depend on _MatMulModule**
- **Nest _MatMulModule inside InterModuleParamReuseModel**
- **Rename most architectures**
- **Revamp WithEmptyOutput**
- **Revamp WithBuffered**
- **Revamp Cifar10Model**
- **Add docstring to FreeParam**
- **Add docstring to NoFreeParam**
- **Add docstring to UnusedParam and rename it to SomeUnusedParam**
- **Add docstring to SomeFrozenParam**
- **Reorder some classes in _architectures**
- **Add missing parametrizations in test_augment_model.py**
- **Remove n_iter param from test_equivalence, fix it to 3**
- **Add STANDARD_PARAMETRIZATIONS and BENCHMARKING_PARAMETRIZATIONS**
- **Improve parametrizations**
- **Change `HandleManager` to `torchjd.RemovableHandle`.**
- **Desambiguate the name of `RemovableHandle`s from torch and torchjd.**
- **Restore documentation of `RemovableHandle.remove` and add documentation of `RemovableHandle`**
- **Fix type hint for handles in augment_model_for_gramian_based_iwrm**
- **Fix docstring of RemovableHandle**
- **Improve formulation of comment in handle usage example**
- **Reduce batch size in InstanceNormResNet18 parametrization**
- **Reduce learning rate in test_equivalence to 1e-7**
- **Revamp WithEmptyOutput**
- **Split speed and unit tests for autogram**
- **Make all subclass of `Weighting[PSDMatrix]` public: * Give corresponding default values to their init * Add them in `__init__.py` * Publish `Weighting`, `Matrix` and `PSDMatrix`**
- **Documentation of `Weighting` now reference the public import.**
- **Make some subclasses of `Weighting[Matrix]` public, excludes NashMTL as it gives different weights if we give it a matrix or its Gramian.**
- **Publish the documentation of all weightings available for autogram**
- **Add the list of available weightings to the docstring of `augment_model_for_gramian_based_iwrm`**
- **Move the documentation of `Weighting` in aggregation**
- **Move the documentation of `Handle` in `augment_module`**
- **Remove the example of de-augmenting the module with the Handle. This example is right below in the doc in the documentation of `RemovableHandle`.**
- **Improve reference to `Weighting` in the docstrings.**
- **Change every `Upgrad().weighting.weighting` to `UPgradWrapper()`**
- **Use tree_map_only instead of tree_map. For now, we don't handle modules that takes or outputs anything else than a PyTree of Tensor but still this is safer.**
- **Publish `WeightedAggregator` and `GramianWeightedAggregator`. This enables showing inheritance of all aggregators and weightings.**
- **Remove `plot_graph.py`, we are not using it anymore, and if needed we know where to find it.**
- **Remove asserts, we are now confident enough.**
- **raise an error if the Gramian is `None`.**
- **assert Gramian is not `None`. It cannot be as it would mean that there are no `target_leafs` in the call to `autograd.grad` in `AutogramScaler`, which would make it fail.**
- **Make GramianAccumulator.gramian be Tensor | None and cast in usage**
- **Change type hint from Tensor | None to Optional[Tensor]**
- **Move autograd functions to _forward_hooks.py**
- **Make wrapper weightings take pref_vector instead of weighting**
- **Add explanation about --runslow flag in CONTRIBUTING.md**
- **Move dict_assertions to utils package**
- **Rename wrapper to weighting**
- **Fix docstring**
- **Update `get_leaf_edges` to explore from `roots`. This ensures that if the returned leafs is non-empty, then they are on the graph of the roots.**
- **Make hooks appropriate Callable classes.**
- **Import `PSDMatrix` and `Weighting` correctly**
- **Improve imports**
- **Improve imports**
- **Move notes about importing from _pytree in the appropriate files.**
- **Add aggregator and weighting parametrization in test_equivalence**
- **Reduce gradient magnitudes**
- **Make parametrization on CAGrad optional depending on whether it is installed**
- **Remove outdated comments**
- **Add Randomness failing test**
- **Add some unit tests to `GramianAccumulator`**
- **Improve docstring of Randomness**
- **Mark Randomness test as expected failure**
- **Make `HookActivator` have the responsibility of handling the activation of hooks.**
- **Make `HookActivator` have the responsibility of handling the activation of hooks.**
- **Make `HookActivator` have the responsibility of handling the activation of hooks.**
- **Use curried generators of tensors rather than the ones from torch.**
- **Unify level of abstraction of EdgeRegistry**
- **Reorder architectures**
- **Force using deterministic algorithms only on CPU tests**
- **Fix randomness better in test_augment_model**
- **Add AlexNet**
- **Add garbage_collect marker**
- **Add some initial tests of EdgeRegistry**
- **Update src/torchjd/_autogram/_hook_activator.py**
- **Use addmm in gramian_accumulator tests.**
- **Remove outdated todo in usage example**
- **Fix warning in augment_model_for_gramian_based_iwrm usage example**
- **Update tolerances of gramian_accumulator tests.**
- **Remove duplicate assert_close**
- **Fix the fix of the tolerances of `assert_close` in `gramian_accumulator`...**
- **Add WithSideEffect as test with expected failure**
- **Rename to make_activable_hook**
- **Remove outdated param description in ModuleHook**
- **Rename `HookActivator` to `ActivableHookFactory` and `make_activable_hook` to `make_hook_activable`.**
- **Rename `hook_activator` to `activable_hook_factory`.**
- **Move `ActivableHookFactory` to `_forward_hooks.py`**
- **Move utils to newly created `_utils` package.**
- **Add SomeUnusedOutput**
- **Add failing tests with models with outputs of ndim 0, 1, 2, 3 and 4**
- **Fix multiplication by weights in AutogramScaler**
- **Fix mse_loss_fn in test utils**
- **Add autogram to the IWRM example.**
- **improve autogram code in IWRM example**
- **Add usage example for Partial JD on IWRM**
- **Speak of autogram in `README.md`**
- **Fix usage of autogram in README**
- **Publish `_autogram` and `_autojac`: * rename packages `_autojac` and `_autorgamd` to `autojac` and `autogram` * remove all import in the package `torchjd`, replace all imports from the one corresponding to the appropriate packages (at this point, all tests pass and docs builds correctly) * add back imports in package `torchjd` with deprecation. (and tested that warning are raised if the import is from torchjd).**
- **Improve docstrings of pacakges.**
- **Rename `augment_model_for_gramian_based_iwrm` to `augment_model_for_iwrm`**
- **Documentation of modules are now specified by the module docstring**
- **Move base classes of `aggregation` to the page `docs/aggregation/index.rst`**
- **Improve documentation: * make `combine` private in `WeightedAggregator` as it should not be documented * add constructor to Aggregator and Weighting to improve the display style.**
- **Change improper link to a class (There use to be a page for Aggregator and now there is none)**
- **Rename `augment_model` to `augment_model_for_iwrm` in the documentation**
- **Improve IWRM examples**
- **Improve IWRM examples**
- **Fix formatting**
- **exchange autogram/autojac in iwrm example**
- **remove sentence on top of codes in IWRM example. Those where technically true sentences but very hard to parse and not bringing much additional information.**
- **Move the documentation of  RemoveableHandle from `augment_model_for_iwrm` to `autogram`**
- **Improve slightly documentation of autogram**
- **Import some types from `collections.abc` rather than from `typing`.**
- **Update src/torchjd/autogram/__init__.py**
- **Update src/torchjd/autogram/__init__.py**
- **Improve deprecations**
- **Add division by 5 in Ndim0Output and Ndim1Output**
- **Explain how to deprecate code in Contributing.md**
- **fix link to autojac in documentation**
- **Remove duplicate deprecation date**
- **Add test for the import error when trying to import something that does not exist from torchjd**
- **Rename make_hook_activable to make_activable_hook**
- **Improve garbage_collect_if_marked**
- **Add division by 5 in InstanceNormResNet18 to fix numerical issue**
- **Fix formatting of partial_jd.rst**
- **Stop talking about Relu's params in partial_jd.rst (there are none)**
- **Add comment in partial_jd.rst**
- **Update test_iwrm in test_rst**
- **Add (failing) test_partial_jd in test_rst.py**
- **Fix partial jd example**
- **Add optional submodule parameter to `augment_model_for_iwrm` (need tests)**
- **Update IWRM example to use `submodules` parameter of `augment_model_for_iwrm`**
- **Revamp iwrm.rst**
- **Remove torchjd. prefix from package names in doc**
- **Refactor documentation of aggregation page**
- **Remove autogram._utils package**
- **Remove aggregator usage examples**
- **Add aggregator and weighting usage examples and their doc test**
- **Change of interface.**
- **Change `test_partial` that still needs to be changed.**
- **Rename to `GramianReverseAccumulator`, also make it responsible only for computing the Gramian**
- **Add test that we get a vector in `compute_gramian`, also add a TODO for the test example which fails.**
- **Add documentation to `GramianReverseAccumulator`**
- **Remove `augment_model_for_iwrm`**
- **adapt documentation from `augment_model_for_iwrm` to `GramianReverseAccumulator`**
- **improve error message**
- **Add test to cover non-vector case for `compute_gramian`.**
- **Force CI to pass, probably not the most elegant solution, need to discuss it.**
- **Remove `ModelHook` and `make_gramian_scaler`**
- **Make (gramian accumulation) nodes activable from outside so that they are only ever activated from `compute_gramian` this means that we can now use the model normally (autograd.backward). Added a test to check that if we do, then the gramian is not computed.**
- **Factorize activators and hook managing.**
- **rename `_forward_hook.py` to `_module_hook_manager.py`**
- **Add/improve docstrings of `ModuleHookManager`**
- **Rename GramianReverseAccumulator to Engine**
- **Remove the possibility to de-augment the modules**
- **Fix test_augment_deaugment_reaugment**
- **WIP on vmapable backward phase.**
- **Remove return for the backward functions, that is not the problem.**
- **Rename whatever the fuck it was to test_engine.py**
- **Merge test_autograd_backward_on_augmented_model with test_autograd_while_modules_are_hooked**
- **Rename some things**
- **Rename _track_modules to _hook_modules**
- **Add failing WithSomeFrozenModule**
- **Fix bug of hooking modules that have no parameter requiring grad.**
- **Add MultiOutputWithFrozenBranch**
- **Fix bug of trying to register the edge of a tensor that does not require grad**
- **Factorize aggregator value tests into test_values.py**
- **Add weighting value tests**
- **Update weighting docstrings**
- **Fix formatting in README.md**
- **Remove explanation about how to aggregate a matrix from README.**
- **Revamp README**
- **Add weighting parametrization to test_partial_autogram**
- **Fix randomness before calling weighting in test_partial_autogram**
- **Clean up test_partial_autogram**
- **Fix engine usage example and doctest**
- **Fix iwrm usage examples and doctests**
- **Fix partial jd example and doctest**
- **Remove todo**
- **Update docs/source/index.rst**
- **Moved the temporary `test_backward_vmap` file to the tests of autogram.**
- **Improve the example in `test_bacward_vmap` to ressemble more the structure of the test that fails when run with the engine. Now the model is exaclty the same, we hook modules with a hook that add a Node, this node's backward is a hand made function in which we defined vmap. This is interesting as this works. The path of differentiation really goes to the `vmap` function as it should. Now I don't understand why this doesn't happen in the engin/module_hook_manager as essentially everything is the same.**
- **Improve the fake example `test_non_batched` in `test_backward_vmap`. Now uses `ModuleHookManager` to hook modules. The backward pass seems to take the vmap route so the problem probably is in the engine.**
- **Add `retain_graph` flag to differentiation. Doens't change anything.**
- **Still works on fake and not on real now everything is essentially the same.**
- **Bug reproduced!**
- **Bug found and fixed: using `grad_outputs` instead of `_grad_outputs` (:**
- **Add limitations to the autogram engine documentation**
- **Add check of compatibility before hooking modules**
- **Fix get_flat_vjp**
- **Improve parametrization of test_partial_autogram**
- **Complexify SimpleBranched architecture**
- **Revamp test_partial_autogram**
- **Fix imports of CAGrad and NashMTL in test_values.py**
- **Improve errors in _check_module_is_compatible**
- **Add tests for _check_module_is_compatible**
- **Add missing docstrings to architectures**
- **Fix test_aggregator_output when nashmtl not installed**
- **Gramians in GramianAccumulator can now be multi dimensional. Need tests**
- **Move `vmap` right after `forward` in `AccumulateJacobian`.**
- **Replace `make_jacobian_accumulator` by `apply_apply_jacobian_accumulator`**
- **Add explanation comments to distinguished between the forward and vmap path in `AccumulateJacobian`**
- **Reduce batch sizes in slow test to make them less likely to run out of memory on my gpu**
- **Add GroupNormMobileNetV3Small**
- **Add SqueezeNet**
- **Add comment about other torchvision models**
- **Add InstanceNormMobileNetV2**
- **Can now handle any shape with ndim <= 2.**
- **Remove `test_non_vector_input_to_compute_gramian`**
- **Rename `get_instance_wise_vjp` to `get_functional_vjp`. We use it both for iwrm and non-batched anyways. It is just a vjp that also accepts the inputs.**
- **Revert "Gramians in GramianAccumulator can now be multi dimensional. Need tests"**
- **Remove the multi-dimensional Gramian handling from `GramianAccumulator` to `Engine`.**
- **Add explanation about not vmaping on the args when there is a non-batched dimension case.**
- **Change `raise` to `assert` as we will anyway remove it eventually. The CI should pass now.**
- **`Engine` now accepts an arbitrary number of dimensions, of which any number can be batched/non-batched.**
- **Fixup determining `non_batched_dims`**
- **Change default value of `batched_dims`**
- **Remove non-batched feature**
- **Fix squeezing issues**
- **Remove nitpick ignore from conf.py**
- **Simplify explanation about deprecation in CONTRIBUTING.md**
- **Remove extra whiteline in imtl_g.rst**
- **Fix extra .backward in basic_usage.rst**
- **Fix engine.rst**
- **Fix intro to IWRM**
- **Improve docstring in _edge_registry.py**
- **Improve docstring in _edge_registry.py**
- **Improve docstring in _edge_registry.py**
- **Improve docstring in _module_hook_manager.py**
- **Improve docstring in _module_hook_manager.py**
- **Improve docstring in _module_hook_manager.py**
- **Fix docstring of test_aggregation.py**
- **Add changelog entry**
- **Update CHANGELOG.md**
- **Update CHANGELOG.md**
- **Update tests/doc/test_aggregation.py**
- **Update src/torchjd/autojac/__init__.py**
- **Update src/torchjd/aggregation/__init__.py**
- **Update docs/source/examples/partial_jd.rst**
- **Update docs/source/examples/iwrm.rst**
- **Update docs/source/examples/iwrm.rst**
- **Update docs/source/examples/iwrm.rst**
- **Update README.md**
- **Update README.md**
- **Update CHANGELOG.md**
- **Fix docstring formatting**
- **Fix renamed file in CONTRIBUTING.md**
- **Fix table formatting in README**
- **Fix autogram IWRM example in README.md**
- **Revert removal of space at the beginning of the documentation**
- **Improve links to autogram engine**
- **Improve explanation of autojac in iwrm.rst**
- **Improve partial_jd.rst**
- **Import directly torchjd.autojac and use autojac.backward in basic usage example**
- **Improve test_edge_registry**
- **Remove parameter excluded from get_leaf_edges**
- **Fix typo**
- **Add type hints of test_gramian_accumulator**
- **Fix import in forward_backwards.py**
- **Improve clarity of usage examples**
- **Renames, documentation and style in test_engine.py**
- **Simplify compute_gramian documentation**
- **Fix typo in error message**
- **Make usage examples match doc tests**
- **Reorder tests in test_rst.py in alphabetical order**
- **Revert changes to README.md**
- **Update `README.md` for autogram**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -104,49 +119,66 @@ using [UPGrad](https://torchjd.org/stable/docs/aggregation/upgrad/).
...
+-     loss = loss_fn(output, target)  # shape [1]
++     losses = loss_fn(output, target)  # shape [16]
+
+      optimizer.zero_grad()
+-     loss.backward()
++     gramian = engine.compute_gramian(losses)  # shape: [16, 16]
++     weights = weighting(gramian)  # shape: [16]
++     losses.backward(weights)
+      optimizer.step()
+```
```
Comment (PierreQuinton) [inline README.md:157]: Will this be the most important example after #411 ?

Comment (ValerianRey) [reply README.md:157]: A bit early to say. Depends on what we want people to do with our library. It may be good to start with this example and to have a more complex one after. To have like:
- MTL with autojac
- IWRM with autogram
- MTL + IWRM with autogram

I think it's always much easier to have the simpler things come first and complexify after, even if what we want people to do is the complex stuff.

This is just the state of the README as it was in #387 btw, nothing new in this PR.

File: README.md [hunk 2]
```diff
@@ -104,49 +119,66 @@ using [UPGrad](https://torchjd.org/stable/docs/aggregation/upgrad/).
...
++ from torchjd.aggregation import UPGradWeighting
+
+  model = Sequential(Linear(10, 5), ReLU(), Linear(5, 3), ReLU(), Linear(3, 1), ReLU())
+
+- loss_fn = MSELoss()
++ loss_fn = MSELoss(reduction="none")
+  optimizer = SGD(model.parameters(), lr=0.1)
+
++ weighting = UPGradWeighting()
++ engine = Engine(model.modules())
```
Comment (PierreQuinton) [inline README.md]: Should we specify the batch_dim?

File: README.md (hunk 1 above)
Comment (PierreQuinton) [reply README.md:157]: I guess we can now answer that? What do you think @ValerianRey ?

File: README.md (hunk 2 above)
Comment (ValerianRey) [reply README.md]: Yes! This parameter is not optional anymore, so this code wouldn't even run! Good catch!

File: README.md (hunk 1 above)
Comment (ValerianRey) [reply README.md:157]: I think we can add IWMTL as a third example in the readme.

**Comment by ValerianRey** [comment]
Need to update to new interface if we merge #451
